packages/swing-store/src/transcriptStore.js

```diff
@@ -0,0 +1,383 @@
+    sqlWriteSpan.run(vatID, 0, 0, initialHash, 1, 0);
+  }
+
+  const sqlGetCurrentTranscriptBounds = db.prepare(`
+    SELECT startPos, endPos, size, hash
+    FROM transcriptSpans
+    WHERE vatID = ? AND current = 1
+  `);
+
+  /**
+   * @param {string} vatID
+   * @returns {{startPos: number, endPos: number, size: number, hash: string}}
+   */
+  function getCurrentTranscriptBounds(vatID) {
+    const bounds = sqlGetCurrentTranscriptBounds.get(vatID);
+    assert(bounds, `no current transcript for ${vatID}`);
+    return bounds;
+  }
+
+  const sqlEndCurrentTranscript = db.prepare(`
+    UPDATE transcriptSpans
+    SET current = 0
+    WHERE current = 1 AND vatID = ?
+  `);
+
+  function rolloverSpan(vatID) {
+    ensureTxn();
+    const { hash, startPos, endPos, size } = getCurrentTranscriptBounds(vatID);
+    noteExport(
+      artifactName({ vatID, startPos }),
+      JSON.stringify({ vatID, hash, startPos, endPos, size }),
+    );
+    sqlEndCurrentTranscript.run(vatID);
+    sqlWriteSpan.run(vatID, endPos, endPos, initialHash, 1, 0);
+  }
+
+  function* getArtifactNames(includeHistorical) {
+    const sql = includeHistorical
+      ? sqlGetAllTranscriptMetadata
+      : sqlGetCurrentTranscriptMetadata;
+    for (const rec of sql.iterate()) {
+      yield artifactName(rec);
+    }
+  }
+
+  const sqlGetTranscriptEndPos = db.prepare(`
+    SELECT endPos
+    FROM transcriptSpans
+    WHERE vatID = ? AND startPos = ?
+  `);
+  sqlGetTranscriptEndPos.pluck(true);
+
+  const sqlReadTranscriptItems = db.prepare(`
+    SELECT item
+    FROM transcriptItems
+    WHERE vatID = ? AND ? <= position AND position < ?
+    ORDER BY position
+  `);
+
+  /**
+   * @param {string} vatID
+   * @param {number} [startPos]
+   */
+  function readTranscript(vatID, startPos) {
+    let endPos;
+    if (startPos === undefined) {
+      ({ startPos, endPos } = getCurrentTranscriptBounds(vatID));
+    } else {
+      assert.typeof(startPos, 'number');
+      endPos = sqlGetTranscriptEndPos.get(vatID, startPos);
+      assert.typeof(
+        endPos,
+        'number',
+        `no transcript for ${vatID} at ${startPos}`,
+      );
+    }
+    !transcriptStatus.get(vatID) ||
+      Fail`can't read transcript ${q(vatID)} because it's already in use`;
+    insistTranscriptPosition(startPos);
+    startPos <= endPos || Fail`${q(startPos)} <= ${q(endPos)}}`;
+
+    function* reader() {
+      for (const { item } of sqlReadTranscriptItems.iterate(
+        vatID,
+        startPos,
+        endPos,
+      )) {
+        transcriptStatus.get(vatID) === 'reading' ||
+          Fail`can't read transcript ${q(vatID)}, it's been closed`;
+        yield item.substring(0, item.length - 1);
+      }
+      transcriptStatus.delete(vatID);
+    }
+    !transcriptStatus.has(vatID) ||
+      Fail`can't read transcript ${q(vatID)} because it's already in use`;
+
+    if (startPos === endPos) {
+      return empty();
+    }
+
+    transcriptStatus.set(vatID, 'reading');
```

Comment:
It's dead to me.  (Soon to you also.)

Reply:
Once upon a time, transcripts were stored as concatenated bytes in one-file-per-vat, and `pos` was an object with both a byte offset and a deliveryNum integer. We needed the byte offset to `seek()` into the file to start replay from a non-initial delivery, and to overwrite the abandoned tail of the transcript if/when we rolled back a delivery. To avoid hopeless confusion if/when an active reader was reading from the middle of the file, and an active writer was appending to the end of the file, the first implementation tracked a `status`: either one writer, or some number of readers, never both.

Moving both the stream store and the kvStore entries (which point into it) into the same SQLite DB means we no longer need to worry about this safety measure: we aren't going to corrupt a file by writing bytes halfway into the middle of an existing file, since there are no files anymore.